src/indexer/path_to_unordered_id.rs

```diff
@@ -0,0 +1,76 @@
+use fnv::FnvHashMap;
+
+#[derive(Default)]
+pub(crate) struct PathToUnorderedId {
+    map: FnvHashMap<String, u32>,
+}
+
+impl PathToUnorderedId {
+    #[inline]
+    pub(crate) fn get_id(&mut self, path: &str) -> u32 {
+        if let Some(id) = self.map.get(path) {
+            return *id;
+        }
+        self.insert_new_path(path)
+    }
+    #[cold]
+    fn insert_new_path(&mut self, path: &str) -> u32 {
+        let next_id = self.map.len() as u32;
+        self.map.insert(path.to_string(), next_id);
+        next_id
+    }
+
+    /// Retuns ids which reflect the lexical order of the paths.
+    ///
+    /// The returned vec can be indexed with the unordered id to get the ordered id.
+    pub(crate) fn unordered_id_to_ordered_id(&self) -> Vec<u32> {
+        let mut sorted_ids: Vec<(&String, &u32)> = self.map.iter().collect::<_>();
+        sorted_ids.sort_unstable_by_key(|(path, _)| *path);
+        let mut result = vec![sorted_ids.len() as u32; sorted_ids.len()];
+        for (ordered, unordered) in sorted_ids.iter().map(|(_k, v)| v).enumerate() {
+            result[**unordered as usize] = ordered as u32;
+        }
+        result
+    }
+
+    /// Retuns ids which reflect the lexical order of the paths.
+    ///
+    /// The returned vec can be indexed with the unordered id to get the ordered id.
+    pub(crate) fn ordered_id_to_path(&self) -> Vec<&String> {
+        let unordered_id_to_ordered_id = self.unordered_id_to_ordered_id();
```

Comment:
I am likely missing something, but wouldn't it be simpler to just do

```rust
let mut paths = self.map.keys().collect::<Vec<_>>();
paths.sort_unstable_key();
paths
```

than to undo what `unordered_id_to_ordered_id` did?

Reply:
(I may also just be confused because the doc comment seems to be copied from the previous function without adjusting it to this one.)

Reply:
yes that's more straightforward